pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java

```diff
@@ -0,0 +1,383 @@
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        replayCursor(future);
+        return future;
+    }
+
+    /**
+     * Read all the entries published from the cursor position until the most recent and update the highest sequence id
+     * from each producer
+     *
+     * @param future
+     *            future to trigger when the replay is complete
+     */
+    private void replayCursor(CompletableFuture<Void> future) {
+        managedCursor.asyncReadEntries(100, new ReadEntriesCallback() {
+            @Override
+            public void readEntriesComplete(List<Entry> entries, Object ctx) {
+
+                for (Entry entry : entries) {
+                    ByteBuf messageMetadataAndPayload = entry.getDataBuffer();
+                    MessageMetadata md = Commands.parseMessageMetadata(messageMetadataAndPayload);
+
+                    String producerName = md.getProducerName();
+                    long sequenceId = md.getSequenceId();
+                    highestSequencedPushed.put(producerName, sequenceId);
+                    highestSequencedPersisted.put(producerName, sequenceId);
+
+                    md.recycle();
+                    messageMetadataAndPayload.release();
+                }
+
+                if (managedCursor.hasMoreEntries()) {
+                    // Read next batch of entries
+                    pulsar.getExecutor().submit(() -> replayCursor(future));
+                } else {
+                    // Done replaying
+                    future.complete(null);
+                }
+            }
+
+            @Override
+            public void readEntriesFailed(ManagedLedgerException exception, Object ctx) {
+                future.completeExceptionally(exception);
+            }
+        }, null);
+    }
+
+    public CompletableFuture<Void> initialize() {
+        // Check whether the dedup cursor was already present
+        for (ManagedCursor cursor : managedLedger.getCursors()) {
+            if (cursor.getName().equals(PersistentTopic.DEDUPLICATION_CURSOR_NAME)) {
+                // Deduplication was enabled before
+                this.status = Status.Recovering;
+                this.managedCursor = cursor;
+                break;
+            }
+        }
+
+        if (status == Status.Recovering) {
+            // Recover the current cursor and then check the configuration
+            return recoverSequenceIdsMap().thenCompose(v -> checkStatus());
+        } else {
+            // No-op
+            return CompletableFuture.completedFuture(null);
+        }
+    }
+
+    /**
+     * Check the status of deduplication. If the configuration has changed, it will enable/disable deduplication,
+     * returning a future to track the completion of the task
+     */
+    public CompletableFuture<Void> checkStatus() {
+        if (status == Status.Recovering || status == Status.Removing) {
```

Comment:
can we move this check at ln#197 after getting `isDeduplicationEnabled()` because sometimes, we have seen (mostly while removing replicator cluster from policies) that broker receives multiple zk-watch event and we have to handle this race condition. Here, `isDeduplicationEnabled()` is an async therefore, it make sense to move this check once we get the result from it (if possible then under synchronized block).

Reply:
Makes sense, also I'll add a mutex in the state mutation